### ISSUE TITLE
optimize hyperparameter tutorial not working

### ISSUE THREAD

**Issue by cakester** [ivadomed/ivadomed#864]
## Issue description
I followed tutorial for optimize hyperparameter in https://ivadomed.org/en/latest/tutorials/automate_training.html

### Current behavior

I got the error message 
```
Traceback (most recent call last):
  File "c:\users\hanna.ma\appdata\local\programs\python\python38\lib\multiprocessing\pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "c:\users\hanna.ma\appdata\local\programs\python\python38\lib\multiprocessing\pool.py", line 48, in mapstar
    return list(map(*args))
  File "c:\users\hanna.ma\appdata\local\programs\python\python38\lib\site-packages\ivadomed\scripts\automate_training.py", line 95, in train_worker
    ivado.run_command(config, thr_increment=thr_incr)
  File "c:\users\hanna.ma\appdata\local\programs\python\python38\lib\site-packages\ivadomed\main.py", line 311, in run_command
    path_output = set_output_path(context)
  File "c:\users\hanna.ma\appdata\local\programs\python\python38\lib\site-packages\ivadomed\main.py", line 195, in set_output_path
    os.makedirs(path_output)
  File "c:\users\hanna.ma\appdata\local\programs\python\python38\lib\os.py", line 223, in makedirs
    mkdir(name, mode)
NotADirectoryError: [WinError 267] The directory name is invalid: "spineGeneric-loss={'name': 'DiceLoss'}"
```

### Installed packages

<details open>
<summary> Output of <code>pip freeze</code> </summary>

```

```

</details>

**Comment by dyt811** [comment]
Can you provide the `config_hyper.json` file you have created during the tutorial? Specifically, how you changed the "loss" key there? 
Given that the `set_output_path` function is working as intended, the error suggests JSON config file is either malformed or passing wrong information to the `set_output_path` command.

**Comment by cakester** [comment]
this is config_hyper.json, exactly like the tutorial
{
    "training_parameters": {
        "batch_size": [2, 64],
        "loss": [
            {"name": "DiceLoss"},
            {"name": "FocalLoss", "gamma": 0.2, "alpha" : 0.5}
        ]
    },
    "default_model": {"depth": [2, 3, 4]}
}

**Comment by mariehbourget** [comment]
I'm working on some hyperparameters optimization as well and I ran into a very similar issue, where some configs would work and others would give me an error. With the same `config_hyper.json` as above, `batch_size` and `depth` training are running properly. However the `loss` ones give me this error:
<details>
<summary> Error </summary>

```
--- Logging error ---
Traceback (most recent call last):
  File "/home/mhbourget/code/ivadomed/ivadomed/scripts/automate_training.py", line 95, in train_worker
    ivado.run_command(config, thr_increment=thr_incr)
  File "/home/mhbourget/code/ivadomed/ivadomed/main.py", line 324, in run_command
    logger.add(log_file)
  File "/home/mhbourget/venv-ivadomed-38c/lib/python3.8/site-packages/loguru/_logger.py", line 762, in add
    wrapped_sink = FileSink(path, **kwargs)
  File "/home/mhbourget/venv-ivadomed-38c/lib/python3.8/site-packages/loguru/_file_sink.py", line 167, in __init__
    self._initialize_file()
  File "/home/mhbourget/venv-ivadomed-38c/lib/python3.8/site-packages/loguru/_file_sink.py", line 186, in _initialize_file
    path = self._prepare_new_path()
  File "/home/mhbourget/venv-ivadomed-38c/lib/python3.8/site-packages/loguru/_file_sink.py", line 179, in _prepare_new_path
    path = self._path.format_map({"time": FileDateFormatter()})
KeyError: "'name'"

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.8/logging/__init__.py", line 1081, in emit
    msg = self.format(record)
  File "/usr/lib/python3.8/logging/__init__.py", line 925, in format
    return fmt.format(record)
  File "/usr/lib/python3.8/logging/__init__.py", line 664, in format
    record.message = record.getMessage()
  File "/usr/lib/python3.8/logging/__init__.py", line 369, in getMessage
    msg = msg % self.args
TypeError: not all arguments converted during string formatting
Call stack:
  File "<string>", line 1, in <module>
  File "/usr/lib/python3.8/multiprocessing/spawn.py", line 116, in spawn_main
    exitcode = _main(fd, parent_sentinel)
  File "/usr/lib/python3.8/multiprocessing/spawn.py", line 129, in _main
    return self._bootstrap(parent_sentinel)
  File "/usr/lib/python3.8/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/usr/lib/python3.8/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python3.8/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/usr/lib/python3.8/multiprocessing/pool.py", line 48, in mapstar
    return list(map(*args))
  File "/home/mhbourget/code/ivadomed/ivadomed/scripts/automate_training.py", line 99, in train_worker
    logging.info("Unexpected error:", sys.exc_info()[0])
Message: 'Unexpected error:'
Arguments: (<class 'KeyError'>,)
--- Logging error ---
Traceback (most recent call last):
  File "/home/mhbourget/code/ivadomed/ivadomed/scripts/automate_training.py", line 95, in train_worker
    ivado.run_command(config, thr_increment=thr_incr)
  File "/home/mhbourget/code/ivadomed/ivadomed/main.py", line 324, in run_command
    logger.add(log_file)
  File "/home/mhbourget/venv-ivadomed-38c/lib/python3.8/site-packages/loguru/_logger.py", line 762, in add
    wrapped_sink = FileSink(path, **kwargs)
  File "/home/mhbourget/venv-ivadomed-38c/lib/python3.8/site-packages/loguru/_file_sink.py", line 167, in __init__
    self._initialize_file()
  File "/home/mhbourget/venv-ivadomed-38c/lib/python3.8/site-packages/loguru/_file_sink.py", line 186, in _initialize_file
    path = self._prepare_new_path()
  File "/home/mhbourget/venv-ivadomed-38c/lib/python3.8/site-packages/loguru/_file_sink.py", line 179, in _prepare_new_path
    path = self._path.format_map({"time": FileDateFormatter()})
KeyError: "'name'"

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.8/logging/__init__.py", line 1081, in emit
    msg = self.format(record)
  File "/usr/lib/python3.8/logging/__init__.py", line 925, in format
    return fmt.format(record)
  File "/usr/lib/python3.8/logging/__init__.py", line 664, in format
    record.message = record.getMessage()
  File "/usr/lib/python3.8/logging/__init__.py", line 369, in getMessage
    msg = msg % self.args
TypeError: not all arguments converted during string formatting
Call stack:
  File "<string>", line 1, in <module>
  File "/usr/lib/python3.8/multiprocessing/spawn.py", line 116, in spawn_main
    exitcode = _main(fd, parent_sentinel)
  File "/usr/lib/python3.8/multiprocessing/spawn.py", line 129, in _main
    return self._bootstrap(parent_sentinel)
  File "/usr/lib/python3.8/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/usr/lib/python3.8/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python3.8/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/usr/lib/python3.8/multiprocessing/pool.py", line 48, in mapstar
    return list(map(*args))
  File "/home/mhbourget/code/ivadomed/ivadomed/scripts/automate_training.py", line 99, in train_worker
    logging.info("Unexpected error:", sys.exc_info()[0])
Message: 'Unexpected error:'
Arguments: (<class 'KeyError'>,)
```
</details>

Although, it is not the same error as @cakester, it seems to be related to the name of the `path_output`. Here https://github.com/ivadomed/ivadomed/blob/93b9e7315565d05f716d2e32feb7e1f1baebe6cb/ivadomed/scripts/automate_training.py#L279 a suffix based on the config (`hyper_option.name`) is added to  `path_output`. In the case of the loss configs, the suffix would be: `-loss={'name': 'DiceLoss'}` and `-loss={"name": "FocalLoss", "gamma": 0.2, "alpha" : 0.5}`.

I think both our errors are due to the illegal characters in the new `path_output` name.
I tried removing or replacing the illegal characters in the name with the following code and it fixes my issue on Ubuntu 20.04:
```
folder_name_suffix = hyper_option.name
folder_name_suffix = folder_name_suffix.translate({ord(i):None for i in '}{ \''})
folder_name_suffix = folder_name_suffix.translate({ord(i):'-' for i in ':=,'})            
new_config["path_output"] = initial_config["path_output"] + folder_name_suffix
```

Notes:
- The same `hyper_option.name` is used [here](https://github.com/ivadomed/ivadomed/blob/93b9e7315565d05f716d2e32feb7e1f1baebe6cb/ivadomed/scripts/automate_training.py#L257) and [here](https://github.com/ivadomed/ivadomed/blob/93b9e7315565d05f716d2e32feb7e1f1baebe6cb/ivadomed/scripts/automate_training.py#L272) in the script.
- II think this naming scheme could cause other errors on some OS when using the `all_combin` option if the path ends up being very long (>256 characters).
